libcxx/include/__flat_map/flat_map.h

```diff
@@ -0,0 +1,1280 @@
+      }
+    } else {
+      return __binary_search_emplace_impl(std::move(__pair)).first;
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI pair<iterator, bool> insert(const value_type& __x) { return emplace(__x); }
+
+  _LIBCPP_HIDE_FROM_ABI pair<iterator, bool> insert(value_type&& __x) { return emplace(std::move(__x)); }
+
+  _LIBCPP_HIDE_FROM_ABI iterator insert(const_iterator __hint, const value_type& __x) {
+    return emplace_hint(__hint, __x);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI iterator insert(const_iterator __hint, value_type&& __x) {
+    return emplace_hint(__hint, std::move(__x));
+  }
+
+  template <class _Pp>
+    requires is_constructible_v<pair<key_type, mapped_type>, _Pp>
+  _LIBCPP_HIDE_FROM_ABI pair<iterator, bool> insert(_Pp&& __x) {
+    return emplace(std::forward<_Pp>(__x));
+  }
+
+  template <class _Pp>
+    requires is_constructible_v<pair<key_type, mapped_type>, _Pp>
+  _LIBCPP_HIDE_FROM_ABI iterator insert(const_iterator __hint, _Pp&& __x) {
+    return emplace_hint(__hint, std::forward<_Pp>(__x));
+  }
+
+  template <input_iterator _InputIterator>
+  _LIBCPP_HIDE_FROM_ABI void insert(_InputIterator __first, _InputIterator __last) {
+    if constexpr (sized_sentinel_for<_InputIterator, _InputIterator>) {
+      __reserve_impl(__last - __first);
+    }
+
+    for (; __first != __last; ++__first) {
+      __binary_search_emplace_impl(value_type(*__first));
+    }
+  }
+
+  template <input_iterator _InputIterator>
+  void insert(sorted_unique_t, _InputIterator __first, _InputIterator __last) {
+    if constexpr (sized_sentinel_for<_InputIterator, _InputIterator>) {
+      __reserve_impl(__last - __first);
+    }
+
+    auto __it = begin();
+    while (__first != __last) {
+      value_type __pair(*__first);
+      auto __end = end();
+      __it       = ranges::lower_bound(__it, __end, __pair.first, __compare_, [](const auto& __p) -> decltype(auto) {
+        return std::get<0>(__p);
+      });
+      if (__it == __end || __compare_(__pair.first, __it->first)) {
+        __it = __emplace_impl(__it, std::move(__pair));
+      }
+      ++__it;
+      ++__first;
+    }
+  }
+
+  template <_ContainerCompatibleRange<value_type> _Range>
+  _LIBCPP_HIDE_FROM_ABI void insert_range(_Range&& __range) {
+    if constexpr (ranges::sized_range<_Range>) {
+      __reserve_impl(ranges::size(__range));
+    }
+
+    auto __last = ranges::end(__range);
+    for (auto __it = ranges::begin(__range); __it != __last; ++__it) {
+      __binary_search_emplace_impl(value_type(*__it));
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI void insert(initializer_list<value_type> __il) { insert(__il.begin(), __il.end()); }
+
+  _LIBCPP_HIDE_FROM_ABI void insert(sorted_unique_t __s, initializer_list<value_type> __il) {
+    insert(__s, __il.begin(), __il.end());
+  }
+
+  _LIBCPP_HIDE_FROM_ABI containers extract() && {
+    auto __guard = std::__make_exception_guard([&]() noexcept { clear() /* noexcept */; });
+    auto __ret   = std::move(__containers_);
```

Comment:
You need to `clear()` unconditionally to handle the case where the containers are in an inconsistent state after being moved-from. Needs a test.

You'll need to do this with some kind of `__scoped_guard` here (and probably elsewhere) since this is not related to exceptions, and `__exception_guard` doesn't rollback anything under `-fno-exceptions`.